### ISSUE TITLE
mipi_pll_div2_clk setting (using .dtsi) does update i.MX8QX clocking tree with the bogus value

### ISSUE THREAD

**Issue by ZoranStojsavljevic** [varigit/linux-imx#12]
Hello Variscite developers,

**_[The Problem Description]_**

According the the iMX8QM reference manual the main PLL for MIPI DSI subsystem is 864 MHz and is not supposed to be internally configurable (fixed clock, 24 MHz x 36). It is explicitly stated in the i.MX8QX RM on the page 447:
8.11.3.2 Subsystem PLLs ==>> Table 8-17. Subsystem PLLs.

The mipi_pll_div2_clk is derived from the main PLL divided by 2, which is actually 432 MHz as expected:

> a72_clk                              0        0        0  1296000000          0     0  50000
>  a53_clk                              0        0        0  1200000000          0     0  50000
>  **_mipi_pll_div2_clk                    2        2        0   432000000          0     0  50000_**
>     mipi0_dsi_rx_esc_clk              0        0        0    72000000          0     0  50000
>     mipi0_dsi_tx_esc_clk              1        1        0    18000000          0     0  50000
>     mipi0_dsi_phy_clk                 1        1        0    27000000          0     0  50000
>     mipi1_dsi_phy_clk                 0        0        0    27000000          0     0  50000

My problem is that mipi_pll_div2_clk (432 MHz clock) is not able to drive the non standard display on the MIPI DSI 0 channel.

**_[Proposed Solution Description]_**

So, I got an idea how to fix this problem using a .dtsi include file. Using arch/arm64/boot/dts/freescale/imx8qm-ss-mipi.dtsi , I decided to change mipi_pll_div2_clk from 432 MHz to 864 MHz.

And here is the actual patch, what is changed:

> diff --git a/arch/arm64/boot/dts/freescale/imx8qm-ss-mipi.dtsi b/arch/arm64/boot/dts/freescale/imx8qm-ss-mipi.dtsi
> index ba166abbab3c..537ffbc2422a 100644
> --- a/arch/arm64/boot/dts/freescale/imx8qm-ss-mipi.dtsi
> +++ b/arch/arm64/boot/dts/freescale/imx8qm-ss-mipi.dtsi
> @@ -15,7 +15,8 @@
>         mipi_pll_div2_clk: clock-mipi-div2-pll {
>                 compatible = "fixed-clock";
>                 #clock-cells = <0>;
>  _[del]         clock-frequency = <432000000>;_
>  _[add]        // clock-frequency = <432000000>;_
>  _[add]        clock-frequency = <864000000>;_
>                  clock-output-names = "mipi_pll_div2_clk";
>         };

**_[The Actual Results]_**

I recompiled the kernel, and loaded it on the Android device I am using. The formal results are shown here:

> a72_clk                              0        0        0  1296000000          0     0  50000
>  a53_clk                              0        0        0  1200000000          0     0  50000
>  **_mipi_pll_div2_clk                    0        0        0   864000000          0     0  50000_** 
>     mipi0_dsi_rx_esc_clk              0        0        0    72000000          0     0  50000
>     mipi0_dsi_tx_esc_clk              0        0        0    18000000          0     0  50000
>     mipi0_dsi_phy_clk                 0        0        0    24000000          0     0  50000
>     mipi1_dsi_phy_clk                 0        0        0    27000000          0     0  50000

But when I tested the display, I discovered **_that the actual mipi_pll_div2_clk did NOT change, and it's real value stayed 432 MHz_**.

As my best guess, the clk_summary command should reflect the real clocks, not a formal changes, introduced from the .dtsi and .dts files?!

This phenomenon comes from the fact that the SCU unit provides the i.MX8QX clocks, and my kludge/assumption is that SCU does NOT communicate clocks back to the i.MX8QX's main unit, should it?

Is there any method to change the mipi_pll_div2_clk on the i.MX8QX silicons, actually NOT reprogramming the SCU unit (even if this is still possible)?

Zoran
_______

**Comment by varigigi** [comment]
Hi Zoran,
as already explained in our customer portal to one of your colleagues (ticket 17004), there's no documentation about the possible way to change it.
According the the iMX8QX reference manual (section 8.11.3.2, table 8-17), the main PLL for MIPI DSI subsystem is 864 MHz and is not supposed to be configurable.
The name "mipi_pll_div2_clk" let me think to the main PLL divided by 2 that is actually 432 MHz as expected.
This is just how the SoC works, there's nothing to be fixed or changed.
Since we are already supporting your colleague in the customer portal, I'm now closing this issue.
BR
Pier

**Comment by ZoranStojsavljevic** [comment]
> The name "mipi_pll_div2_clk" let me think to the main PLL
> divided by 2 that is actually 432 MHz as expected.

Hello Pier,

I object and do not agree with this sentence, and the statement. Since I got 864 MHz for the  mipi_pll_div2_clk. So, in the kernel run time the silicon must be able to output the correct freq, which in this case it does not.

So, we here are dealing with the bug. Since there is NO feedback between SCU unit and the main unit, my best guess.

Yes, the main PLL is fixed clock (said in the documentation, page 447), but nothing is said about mipi_pll_div2_clk in the i.MX8QX reference manual.

I think you need to provide deeper explanation about what is the connection between main PLL and mipi_pll_div2_clk, since the name is intuitive, but nothing more is said about these relations. mipi_pll_div2_clk is not even mentioned in the i.MX8QX reference manual, not even once on 9111 pages of it.

Don't you think, Pier, that this fact is a very strange fact?!

BR,
Zoran
_______

**Comment by varigigi** [comment]
@ZoranStojsavljevic,
IMHO, the best way to get deep details about NXP silicon is to ask NXP.
I opened a dedicated [thread](https://community.nxp.com/t5/i-MX-Processors/mipi-pll-div2-clk-setting/m-p/1477564) in NXP public forum.
Please feel free to contribute.
Thanks
BR
Pier

**Comment by ZoranStojsavljevic** [comment]
Thank you, Pier. Appreciated.

Zoran
_______

**Comment by ZoranStojsavljevic** [comment]
Hello Pier,

Actually, I found the answer on this question. To make it short (after many calculations about the maximum display resolution for MIPI DSI, and finding the correct i.MX8 Manual: i.MX 8QuadMax Applications Processor Reference Manual, Rev. 0, 9/2021), here is the essence.

The question: **_Given the mipi_pll_div2_clk at 432 MHz, which is the maximum display resolution for MIPI DSI ?_**

The answer from the i.MX 8QuadMax Applications Processor Reference Manual, Rev. 0, 9/2021 on page 5208 (chapter 15.3.2.5 DPHY PLL), it states explicitly: _**The VCO maximum output frequency is 800 MHz**_.

In other words, there is a theoretical maximum of 864 Mbit/s per data lane, which for 4 lanes gives maximum throughput of 864 Mbit/s x 4 = 3456 Mbit/s.

But, unfortunately, it is limited (by technology used) to the 800 Mbit/s x 4 = 3200 Mbit/s. The arbitrary display resolution can be used to calculate throughput.

Please, do NOT post this as answer to the NXP topic, since I would like for some time to see if anybody from NXP will take another possible explanation or route, if any???

Thank you,
Zoran
_______
